cmd/clusterctl/pkg/client/cluster/installer.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/repository"
+)
+
+// ProviderInstallerService defined methods for enforcing consistency rules for provider installation.
+type ProviderInstallerService interface {
+	// Add a provider to the install queue.
+	// NB. By deferring the installation, the installer service can perform validation of the target state of the management cluster
+	// before actually starting the installation of new providers.
+	Add(repository.Components, bool) error
+
+	// Perform the installation of the providers ready in the install queue.
+	Install() ([]repository.Components, error)
+}
+
+// installerService implements ProviderInstallerService
+type installerService struct {
+	proxy              Proxy
+	providerComponents ComponentsClient
+	providerInstaller  InventoryClient
+	installQueue       []repository.Components
+}
+
+var _ ProviderInstallerService = &installerService{}
+
+func (i *installerService) Add(components repository.Components, force bool) error {
+	if err := i.providerInstaller.Validate(components.Metadata()); err != nil {
+		if !force {
+			return errors.Wrapf(err, "Installing provider %q can lead to a non functioning management cluster (you can use --force to ignore this error).", components.Name())
+		}
+	}
+
+	i.installQueue = append(i.installQueue, components)
+	return nil
+}
+
+func (i *installerService) Install() ([]repository.Components, error) {
+	ret := make([]repository.Components, len(i.installQueue))
+	for c, components := range i.installQueue {
+		klog.V(3).Infof("Installing provider %s/%s:%s", components.TargetNamespace(), components.Name(), components.Version())
+
+		// create the provider
+		err := i.providerComponents.Create(components)
+		if err != nil {
```

Comment:
How do we plan to handle clean up in case something goes wrong? A simple rollback scheme should be sufficient IMO.

Reply:
what do you mean with a "simple rollback scheme"? 
From my test, the simple create or update logic implemented in Create are ok to cover most of the cases, so my current orientation is to not over-engineer in this first iteration...

Reply:
If `i.providerComponents.Create(components)` were throw an error in the middle -- for example, bootstrap components installed successfully but infra components install failed. There will be some "dangling" state left behind. 
We can leave it to the user to manually cleanup that state or provide some means to undo what has been installed. 
I realize this will complicate matters and I'm fine with it not being included.

Reply:
Once we will have `clusterctl delete` implemented, we can consider to automatically trigger delete if something goes wrong